File: katalon-prep/katalon-prep-python/src/prep/distributed/resilience.py
"""Module 06 — retry / jitter / circuit breaker / rate limit / idempotency.

Doi xung voi `06-distributed-resilience` + `08-system-design` ben Java. Cung thuat toan,
nhung viet lai bang Python de ban tra loi duoc **bang ngon ngu nao ho hoi**.

Nguyen tac cua ca file: **khong goi `time.time()` hay `random.random()` truc tiep.**
Dong ho va bo sinh ngau nhien deu duoc TIEM VAO. Do la ly do duy nhat khien nhung thu
nay test tat dinh duoc — va cung la cau tra loi cho "lam sao test duoc retry/timeout?".
"""

from __future__ import annotations

import hashlib
import time
from collections import defaultdict
from collections.abc import Callable, Iterable, Sequence
from dataclasses import dataclass, field
from enum import Enum
class NonRetryable(Exception):
    """Loi VINH VIEN: 400, 401, 404, schema sai.

    Retry mot loi vinh vien khong bao gio thanh cong — chi ton thoi gian va lam nghen
    hang doi. Phan biet duoc loi tam thoi va loi vinh vien la nua gia tri cua retry.
    """


class Backoff(Enum):
    """Bon chien luoc, xep theo do "an toan cho he thong phia sau"."""

    FIXED = "fixed"
    EXPONENTIAL = "exponential"
    EXPONENTIAL_FULL_JITTER = "exponential_full_jitter"
    EXPONENTIAL_EQUAL_JITTER = "exponential_equal_jitter"


@dataclass(frozen=True, slots=True)
class RetryPolicy:
    max_attempts: int = 3
    base_delay: float = 0.1
    cap: float = 30.0
    strategy: Backoff = Backoff.EXPONENTIAL_FULL_JITTER
    # Tiem bo sinh ngau nhien -> test tat dinh. Mac dinh dung `random.random`.
    rand: Callable[[], float] = field(default=lambda: 0.5, repr=False)

    def __post_init__(self) -> None:
        if self.max_attempts < 1:
            raise ValueError("max_attempts phai >= 1")
        if self.base_delay <= 0:
            raise ValueError("base_delay phai > 0")

    def delay_for(self, attempt: int) -> float:
        """`attempt` tinh tu 0.

        `cap` la thu hay bi quen: khong co no thi lan thu 20 doi 2^20 * 0.1s = 29 gio.
        """
        exp = min(self.base_delay * (2**attempt), self.cap)
        match self.strategy:
            case Backoff.FIXED:
                return self.base_delay
            case Backoff.EXPONENTIAL:
                # KHONG jitter -> moi client sau su co cung thuc day mot luc
                # -> "thundering herd" -> service vua hoi phuc lai gay lai.
                return exp
            case Backoff.EXPONENTIAL_FULL_JITTER:
                # AWS khuyen dung: rai deu tren [0, exp]. Tan xa tot nhat.
                return exp * self.rand()
            case Backoff.EXPONENTIAL_EQUAL_JITTER:
                # Nua co dinh + nua ngau nhien: van tan xa nhung dam bao mot khoang
                # cho toi thieu. Hop khi backend can thoi gian hoi suc chac chan.
                return exp / 2 + (exp / 2) * self.rand()


def retry[T](
    fn: Callable[[], T],
    policy: RetryPolicy | None = None,
    sleep: Callable[[float], None] = time.sleep,
) -> T:
    """Thuc thi co retry. `sleep` tiem vao de test khong phai cho that."""
    p = policy or RetryPolicy()
    last: BaseException | None = None
    for attempt in range(p.max_attempts):
        try:
            return fn()
        except NonRetryable:
            raise  # loi vinh vien: dung ngay, khong lang phi luot retry
        except Exception as ex:
            last = ex
            if attempt < p.max_attempts - 1:
                sleep(p.delay_for(attempt))
    raise RuntimeError(f"that bai sau {p.max_attempts} lan") from last


class BreakerState(Enum):
    CLOSED = "closed"
    OPEN = "open"
    HALF_OPEN = "half_open"


class CircuitBreaker:
    """Ba pha. Khac retry o cho co ban: retry giup MOT request, breaker bao ve CA HE THONG.

    Vi sao can ca hai: retry mot minh lam su co NANG THEM — backend dang qua tai thi
    retry nhan luu luong len gap 3. Breaker ngat han luong di khi da ro la hong.

    HALF_OPEN la pha quan trong nhat va hay bi lam sai: chi cho **mot** request thu di
    qua. Cho ca lo di qua thi backend vua ngoi day lai bi dap chet lan nua.
    """

    def __init__(
        self,
        failure_threshold: int = 5,
        reset_timeout: float = 10.0,
        success_threshold: int = 2,
        clock: Callable[[], float] = time.monotonic,
    ) -> None:
        if failure_threshold < 1 or success_threshold < 1:
            raise ValueError("nguong phai >= 1")
        self._failure_threshold = failure_threshold
        self._reset_timeout = reset_timeout
        self._success_threshold = success_threshold
        self._clock = clock
        self._state = BreakerState.CLOSED
        self._failures = 0
        self._successes = 0
        self._opened_at = 0.0
        self._probe_in_flight = False

    @property
    def state(self) -> BreakerState:
        if self._state is BreakerState.OPEN and self._clock() - self._opened_at >= self._reset_timeout:
            self._state = BreakerState.HALF_OPEN
            self._successes = 0
            self._probe_in_flight = False
        return self._state

    def call[T](self, fn: Callable[[], T]) -> T:
        state = self.state
        if state is BreakerState.OPEN:
            raise BreakerOpen("mach dang ngat - khong goi backend")
        if state is BreakerState.HALF_OPEN:
            if self._probe_in_flight:
                raise BreakerOpen("dang thu mot request - cac request khac phai cho")
            self._probe_in_flight = True
        try:
            result = fn()
        except Exception:
            self._on_failure()
            raise
        else:
            self._on_success()
            return result

    def _on_success(self) -> None:
        self._probe_in_flight = False
        if self._state is BreakerState.HALF_OPEN:
            self._successes += 1
            if self._successes >= self._success_threshold:
                self._state = BreakerState.CLOSED
                self._failures = 0
        else:
            self._failures = 0

    def _on_failure(self) -> None:
        self._probe_in_flight = False
        if self._state is BreakerState.HALF_OPEN:
            # Mot lan hong o HALF_OPEN la mo lai NGAY, khong doi du nguong.
            self._state = BreakerState.OPEN
            self._opened_at = self._clock()
            return
        self._failures += 1
        if self._failures >= self._failure_threshold:
            self._state = BreakerState.OPEN
            self._opened_at = self._clock()


class BreakerOpen(Exception):
    pass


class TokenBucket:
    """Rate limit cho phep BUC ngan han — dung cho API cua san pham.

    So voi fixed window: fixed window co bug bien cua so. Gioi han 100/phut thi client
    gui 100 luc 00:59 va 100 luc 01:00 -> 200 request trong 1 giay ma van "dung luat".
    Token bucket khong co lo do vi no lien tuc.

    Nap token LUOI (lazy): khong co timer nen, chi tinh lai khi co request. Nghia la
    10.000 tenant chi ton 10.000 dong so, khong ton 10.000 timer.
    """

    def __init__(
        self, capacity: int, refill_per_second: float, clock: Callable[[], float] = time.monotonic
    ) -> None:
        if capacity < 1 or refill_per_second <= 0:
            raise ValueError("capacity >= 1 va refill_per_second > 0")
        self._capacity = capacity
        self._refill = refill_per_second
        self._clock = clock
        self._tokens = float(capacity)
        self._last = clock()

    @property
    def tokens(self) -> float:
        self._top_up()
        return self._tokens

    def _top_up(self) -> None:
        now = self._clock()
        elapsed = now - self._last
        if elapsed > 0:
            self._tokens = min(self._capacity, self._tokens + elapsed * self._refill)
            self._last = now

    def try_acquire(self, n: int = 1) -> bool:
        self._top_up()
        if self._tokens >= n:
            self._tokens -= n
            return True
        return False


class PerTenantRateLimiter:
    """Moi tenant mot bucket rieng — cach ly nhieu on.

    Mot bucket dung chung nghia la mot khach chay 10.000 test se lam moi khach khac bi
    tu choi. Trong SaaS thi day la su co ve HOP DONG, khong phai ve ky thuat.
    """

    def __init__(self, capacity: int, refill_per_second: float, clock: Callable[[], float] = time.monotonic) -> None:
        self._capacity = capacity
        self._refill = refill_per_second
        self._clock = clock
        self._buckets: dict[str, TokenBucket] = {}

    def try_acquire(self, tenant: str, n: int = 1) -> bool:
        bucket = self._buckets.get(tenant)
        if bucket is None:
            bucket = TokenBucket(self._capacity, self._refill, self._clock)
            self._buckets[tenant] = bucket
        return bucket.try_acquire(n)

    def __len__(self) -> int:
        return len(self._buckets)


@dataclass
class IdempotencyStore:
    """Chan xu ly trung khi consumer nhan lai message (at-least-once).

    BAY quan trong: TTL. Dat qua ngan thi mot message giao lai muon van bi xu ly hai lan
    -> mat y nghia. Dat qua dai thi bo nho phinh vo han. Phai chon theo do tre TOI DA
    cua he thong (thoi gian giu cua Kafka + so lan retry x backoff), khong chon theo cam tinh.
    """

    ttl: float = 3600.0
    clock: Callable[[], float] = time.monotonic
    _seen: dict[str, float] = field(default_factory=dict, repr=False)

    def mark_if_new(self, event_id: str) -> bool:
        """True neu day la lan dau -> duoc phep xu ly."""
        self._evict()
        if event_id in self._seen:
            return False
        self._seen[event_id] = self.clock()
        return True

    def _evict(self) -> None:
        now = self.clock()
        expired = [k for k, at in self._seen.items() if now - at >= self.ttl]
        for k in expired:
            del self._seen[k]

    def __len__(self) -> int:
        self._evict()
        return len(self._seen)


# ---------------------------------------------------------------------------
# Phan chia partition — noi thang sang bai data skew cua Spark o module 09
# ---------------------------------------------------------------------------


def stable_partition(key: str, partitions: int) -> int:
    """Hash ON DINH giua cac process. **Khong duoc dung `hash()` cua Python.**

    `hash()` cua str duoc lam ngau nhien moi lan khoi dong (PYTHONHASHSEED, chong DoS).
    Dung no de chia partition thi sau khi restart consumer, cung mot key roi vao
    partition khac -> mat sach dam bao thu tu, va bug chi hien sau khi deploy.
    Test o `test_00_refresher.py` chung minh dieu nay bang hai process that.

    Java khong co van de nay (`String.hashCode` co dinh trong spec) — nhung Java LAI co
    van de khac: `hashCode` co the AM, nen phai `Math.floorMod`. Python `%` da floor san.
    Hai ngon ngu, hai cai bay nguoc nhau o cung mot dong code.
    """
    if partitions < 1:
        raise ValueError("partitions phai >= 1")
    digest = hashlib.blake2b(key.encode("utf-8"), digest_size=8).digest()
    return int.from_bytes(digest, "big") % partitions


def partition_sizes(keys: Iterable[str], partitions: int) -> list[int]:
    sizes = [0] * partitions
    for k in keys:
        sizes[stable_partition(k, partitions)] += 1
    return sizes


def skew_ratio(sizes: Sequence[int]) -> float:
    """partition lon nhat / trung binh. 1.0 = deu tuyet doi.

    Day la chi so ban se dung lai NGUYEN VEN o module 09 (Spark): "hot partition" cua
    Kafka va "data skew" cua Spark la **cung mot van de** — mot key chiem qua nhieu
    du lieu — va cung mot cach chua: salting.
    """
    if not sizes:
        return 1.0
    total = sum(sizes)
    if total == 0:
        return 1.0
    return max(sizes) / (total / len(sizes))


def salted_partition(key: str, partitions: int, salt_buckets: int, counter: int) -> int:
    """Chua hot partition bang cach rai mot key nong ra `salt_buckets` gia-key.

    Cai gia phai tra, va PHAI noi ra khi tra loi phong van: **mat dam bao thu tu cho
    key do**. Chi lam duoc khi thu tu trong pham vi key khong quan trong, hoac khi
    ban co the sap xep lai o buoc sau. Doi thu tu lay thong luong.

    ------------------------------------------------------------------------
    BAY QUAN TRONG — do that tren 8 partition, 1 key chiem 90% cua 10.000 ban ghi:

        salt_buckets =   8  (1x so partition)  -> skew 2.80x   (gan nhu vo dung)
        salt_buckets =  16  (2x)               -> skew 3.70x   (**TE HON** ca 1x!)
        salt_buckets =  32  (4x)               -> skew 1.90x
        salt_buckets =  64  (8x)               -> skew 1.45x
        salt_buckets = 128  (16x)              -> skew 1.22x
        salt_buckets = 256  (32x)              -> skew 1.11x

    Ly do: ban chi doi mot key nong thanh N key vua — nhung N key do LAI hash
    khong deu vao cac partition (bai toan sinh nhat). Salt it thi may rui quyet
    dinh, va co truong hop nhieu salt hon lai lech hon.

    **Quy tac:** `salt_buckets >= 10x so partition`. Rat nhieu bai huong dan chi
    ghi "them salt la xong" ma khong noi cho nay — noi duoc con so nay trong phong
    van la khac biet giua "co doc ve salting" va "da tung lam that".

    Neu ban KHONG can thu tu chut nao thi dung `round_robin_partition` ben duoi —
    no san phang tuyet doi va khong phu thuoc may rui.
    """
    if salt_buckets < 1:
        raise ValueError("salt_buckets phai >= 1")
    return stable_partition(f"{key}#{counter % salt_buckets}", partitions)


def round_robin_partition(partitions: int, counter: int) -> int:
    """Rai deu TUYET DOI, khong qua hash.

    Danh doi manh hon salting: mat hoan toan lien he key -> partition, nen khong
    con dinh tuyen duoc va khong con thu tu nao ca. Dung khi ban ghi doc lap that su
    (vi du: gui ket qua tung test case le, moi ban ghi tu chua du thong tin).
    """
    if partitions < 1:
        raise ValueError("partitions phai >= 1")
    return counter % partitions


def hot_key_report(keys: Iterable[str], top: int = 3) -> list[tuple[str, int]]:
    counts: defaultdict[str, int] = defaultdict(int)
    for k in keys:
        counts[k] += 1
    return sorted(counts.items(), key=lambda kv: (-kv[1], kv[0]))[:top]

File: katalon-prep/katalon-prep-java/08-system-design/src/main/java/com/prep/sysdesign/Resilience.java
package com.prep.sysdesign;

import java.time.Duration;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.Map;
import java.util.Set;
import java.util.function.LongSupplier;
import java.util.function.Supplier;

/**
 * SYSTEM DESIGN — ba primitive xuat hien trong gan nhu moi bai design: rate limiter,
 * circuit breaker, va idempotency store.
 *
 * <p><b>Luu y thiet ke chung:</b> ca ba deu nhan {@link LongSupplier} lam nguon thoi gian thay vi
 * goi {@code System.currentTimeMillis()} truc tiep. Do la ly do chung test duoc ma khong can
 * {@code Thread.sleep}: test bơm thoi gian gia. Neu ban lam nguoc lai, test se cham va flaky —
 * day la mot trong nhung quyet dinh testability quan trong nhat khi viet code co yeu to thoi gian.
 */
public final class Resilience {

    private Resilience() {}

    // ==================================================================
    // TOKEN BUCKET RATE LIMITER
    // ==================================================================

    /**
     * Token bucket — quota theo tenant.
     *
     * <p><b>Vi sao token bucket ma khong phai fixed window?</b> Fixed window co bug "bien cua so":
     * gioi han 100 req/phut thi client co the gui 100 req luc 00:59 va 100 req luc 01:00 = 200 req
     * trong 2 giay. Token bucket khong co bien nen khong co bug do.
     *
     * <p><b>Vi sao token bucket ma khong phai leaky bucket?</b> Token bucket cho phep BURST (tich
     * luy token khi ranh roi dung mot lan) — dung voi hanh vi that cua CI: im lang ca dem roi
     * 500 test cung luc luc 9 gio sang. Leaky bucket lam phang tuyet doi, se lam CI cham vo ly.
     *
     * <p>Danh doi cua burst: downstream phai chiu duoc burst = capacity. Nen capacity la mot
     * quyet dinh ve DUNG LUONG, khong phai mot con so tuy y.
     */
    public static final class TokenBucket {

        private final long capacity;
        private final double refillPerMilli;
        private final LongSupplier clock;
        private double tokens;
        private long lastRefillAt;

        /**
         * @param capacity so token toi da (= burst toi da cho phep)
         * @param refillPerSecond so token nap lai moi giay (= throughput on dinh)
         * @param clock nguon thoi gian millis; truyen fake trong test
         */
        public TokenBucket(long capacity, double refillPerSecond, LongSupplier clock) {
            if (capacity < 1) {
                throw new IllegalArgumentException("capacity must be >= 1");
            }
            if (refillPerSecond <= 0) {
                throw new IllegalArgumentException("refillPerSecond must be > 0");
            }
            this.capacity = capacity;
            this.refillPerMilli = refillPerSecond / 1000.0;
            this.clock = clock;
            this.tokens = capacity; // bat dau day -> cho phep burst ngay
            this.lastRefillAt = clock.getAsLong();
        }

        /** @return true neu duoc phep, false neu vuot quota */
        public synchronized boolean tryAcquire(int permits) {
            if (permits < 1) {
                throw new IllegalArgumentException("permits must be >= 1");
            }
            refill();
            if (tokens >= permits) {
                tokens -= permits;
                return true;
            }
            return false;
        }

        public boolean tryAcquire() {
            return tryAcquire(1);
        }

        /**
         * Nap lai token theo thoi gian da troi qua — LAZY, khong can background thread.
         *
         * <p>Day la meo quan trong: mot rate limiter co timer nen dat gap nhieu lan va kho scale
         * (1 trieu tenant = 1 trieu timer). Tinh lai luc doc thi chi phi bang 0 khi khong ai goi.
         */
        private void refill() {
            long now = clock.getAsLong();
            long elapsed = now - lastRefillAt;
            if (elapsed <= 0) {
                return; // dong ho khong lui — chong bug khi NTP dieu chinh nguoc
            }
            tokens = Math.min(capacity, tokens + elapsed * refillPerMilli);
            lastRefillAt = now;
        }

        public synchronized long availableTokens() {
            refill();
            return (long) tokens;
        }
    }

    /** Rate limiter theo tenant — moi tenant mot bucket rieng. */
    public static final class PerTenantRateLimiter {
        private final Map<String, TokenBucket> buckets = new LinkedHashMap<>();
        private final long defaultCapacity;
        private final double defaultRefillPerSecond;
        private final LongSupplier clock;

        public PerTenantRateLimiter(long defaultCapacity, double defaultRefillPerSecond, LongSupplier clock) {
            this.defaultCapacity = defaultCapacity;
            this.defaultRefillPerSecond = defaultRefillPerSecond;
            this.clock = clock;
        }

        public boolean tryAcquire(String tenantId) {
            // computeIfAbsent: bucket duoc tao lazy. Trong production can eviction (Caffeine voi
            // expireAfterAccess) neu khong map se phinh vo han theo so tenant tung goi.
            return buckets
                    .computeIfAbsent(tenantId, k -> new TokenBucket(defaultCapacity, defaultRefillPerSecond, clock))
                    .tryAcquire();
        }

        public int trackedTenants() {
            return buckets.size();
        }
    }

    // ==================================================================
    // CIRCUIT BREAKER
    // ==================================================================

    public enum State {
        /** Binh thuong, cho request di qua. */
        CLOSED,
        /** Downstream dang chet, CHAN het request de no co thoi gian hoi phuc. */
        OPEN,
        /** Thu mot it request de xem downstream da song lai chua. */
        HALF_OPEN
    }

    /** Nem khi circuit dang OPEN — request bi chan truoc khi cham downstream. */
    public static final class CircuitOpenException extends RuntimeException {
        public CircuitOpenException(String message) {
            super(message);
        }
    }

    /**
     * Circuit breaker — may trang thai 3 pha.
     *
     * <p><b>Vi sao can:</b> khi downstream chet, retry lam moi thu TE HON — ban gui gap 3 luu luong
     * vao mot service dang qua tai, va giu thread cua chinh minh de cho timeout. Circuit breaker
     * "that bai nhanh" de bao ve CA HAI phia.
     *
     * <p><b>Vi sao can HALF_OPEN:</b> neu chi co CLOSED/OPEN thi khi het thoi gian cho, TOAN BO
     * luu luong dap vao downstream cung luc va giet no lan nua (thundering herd). HALF_OPEN chi cho
     * mot so request tham do di qua.
     *
     * <p><b>Trade-off:</b> circuit breaker se chan ca nhung request LE RA thanh cong (false
     * positive) trong thoi gian OPEN. Doi lai, downstream co co hoi hoi phuc. Nguong va thoi gian
     * cho la thu phai tune theo so lieu that, khong phai lay mac dinh.
     */
    public static final class CircuitBreaker {

        private final int failureThreshold;
        private final int successThresholdInHalfOpen;
        private final Duration openDuration;
        private final LongSupplier clock;

        private State state = State.CLOSED;
        private int consecutiveFailures;
        private int consecutiveSuccessesInHalfOpen;
        private long openedAt;
        private long rejectedCount;

        public CircuitBreaker(
                int failureThreshold,
                int successThresholdInHalfOpen,
                Duration openDuration,
                LongSupplier clock) {
            if (failureThreshold < 1 || successThresholdInHalfOpen < 1) {
                throw new IllegalArgumentException("thresholds must be >= 1");
            }
            if (openDuration == null || openDuration.isNegative() || openDuration.isZero()) {
                throw new IllegalArgumentException("openDuration must be positive");
            }
            this.failureThreshold = failureThreshold;
            this.successThresholdInHalfOpen = successThresholdInHalfOpen;
            this.openDuration = openDuration;
            this.clock = clock;
        }

        public synchronized <T> T call(Supplier<T> action) {
            transitionIfCoolDownElapsed();

            if (state == State.OPEN) {
                rejectedCount++;
                throw new CircuitOpenException("circuit is OPEN, failing fast without calling downstream");
            }

            try {
                T result = action.get();
                onSuccess();
                return result;
            } catch (RuntimeException ex) {
                onFailure();
                throw ex;
            }
        }

        /** Het thoi gian cho -> OPEN chuyen sang HALF_OPEN de tham do. */
        private void transitionIfCoolDownElapsed() {
            if (state == State.OPEN && clock.getAsLong() - openedAt >= openDuration.toMillis()) {
                state = State.HALF_OPEN;
                consecutiveSuccessesInHalfOpen = 0;
            }
        }

        private void onSuccess() {
            if (state == State.HALF_OPEN) {
                consecutiveSuccessesInHalfOpen++;
                if (consecutiveSuccessesInHalfOpen >= successThresholdInHalfOpen) {
                    state = State.CLOSED;
                    consecutiveFailures = 0;
                }
                return;
            }
            consecutiveFailures = 0;
        }

        private void onFailure() {
            // Mot loi duy nhat trong HALF_OPEN la du de mo lai — downstream ro rang chua san sang.
            if (state == State.HALF_OPEN) {
                open();
                return;
            }
            consecutiveFailures++;
            if (consecutiveFailures >= failureThreshold) {
                open();
            }
        }

        private void open() {
            state = State.OPEN;
            openedAt = clock.getAsLong();
            consecutiveSuccessesInHalfOpen = 0;
        }

        public synchronized State state() {
            transitionIfCoolDownElapsed();
            return state;
        }

        /** So request bi chan — metric phai co, neu khong ban khong biet breaker dang hoat dong. */
        public synchronized long rejectedCount() {
            return rejectedCount;
        }
    }

    // ==================================================================
    // IDEMPOTENCY STORE
    // ==================================================================

    /**
     * Chong xu ly trung — thanh phan bat buoc cua kien truc "at-least-once + idempotent consumer".
     *
     * <p>Kafka/SQS deu la at-least-once: mot message CO THE den hai lan (producer retry, consumer
     * rebalance, commit offset that bai). Exactly-once thuc su rat dat va phuc tap. Cach re va dung
     * la: chap nhan trung, roi lam cho viec xu ly LAP LAI KHONG GAY HAI.
     *
     * <p>Trong DB that, day la {@code INSERT ... ON CONFLICT (event_id) DO NOTHING} — xem Lab 03
     * cua module 05. Ban in-memory nay de test logic va de noi ro vai TTL.
     *
     * <p><b>Vi sao phai co TTL:</b> khong the giu event_id mai mai (het bo nho / het disk). TTL
     * phai LON HON thoi gian retry toi da cua producer. Neu TTL nho hon, mot retry muon se bi coi
     * la event moi va duoc xu ly lan hai — bug rat kho tim vi chi xay ra khi he thong dang co su co.
     */
    public static final class IdempotencyStore {

        private final Map<String, Long> seenAt = new LinkedHashMap<>();
        private final Duration ttl;
        private final LongSupplier clock;

        public IdempotencyStore(Duration ttl, LongSupplier clock) {
            if (ttl == null || ttl.isNegative() || ttl.isZero()) {
                throw new IllegalArgumentException("ttl must be positive");
            }
            this.ttl = ttl;
            this.clock = clock;
        }

        /**
         * @return true neu day la lan DAU tien thay eventId (nen xu ly);
         *     false neu da thay roi (nen bo qua)
         */
        public synchronized boolean markIfFirstTime(String eventId) {
            evictExpired();
            Long previous = seenAt.putIfAbsent(eventId, clock.getAsLong());
            return previous == null;
        }

        private void evictExpired() {
            long cutoff = clock.getAsLong() - ttl.toMillis();
            seenAt.entrySet().removeIf(entry -> entry.getValue() < cutoff);
        }

        public synchronized int size() {
            evictExpired();
            return seenAt.size();
        }
    }

    /**
     * Consumer at-least-once co idempotency — mo phong dung luong xu ly message.
     *
     * <p>Chu y {@code processedOrder}: no chung minh cac side effect chi xay ra MOT lan cho moi
     * eventId, ke ca khi message den nhieu lan.
     */
    public static final class IdempotentConsumer {
        private final IdempotencyStore store;
        private final Set<String> processedOrder = new LinkedHashSet<>();
        private int duplicatesSkipped;

        public IdempotentConsumer(IdempotencyStore store) {
            this.store = store;
        }

        public void consume(String eventId) {
            if (!store.markIfFirstTime(eventId)) {
                duplicatesSkipped++;
                return;
            }
            processedOrder.add(eventId);
        }

        public Set<String> processed() {
            return Set.copyOf(processedOrder);
        }

        public int duplicatesSkipped() {
            return duplicatesSkipped;
        }
    }
}
